Discuss limitations on types based on field def (#2396)
httpwg/http-extensions: draft-ietf-httpbis-sfbis.md

</artwork>
</blockquote>

> Note that a field is defined by referencing a specific RFC for Structured Fields, and the types available for use in its value are limited to those defined in that RFC. For example, a field whose definition references this document can have a value that uses the Date type ({{date}}), whereas a field whose definition references RFC8941 cannot, because it will be treated as invalid (and likely discarded) by implementations of that specification.

Reviewer: ```suggestion
Note that a field is defined by referencing a specific RFC for Structured Fields, and the types available for use in its value are limited to those defined in that RFC. For example, a field whose definition references this document can have a value that uses the Date type ({{date}}), whereas a field whose definition references RFC 8941 cannot, because it will be treated as invalid (and likely discarded) by implementations of that specification.
```

Reviewer: "likely" is a bit weak; not discarding would be a spec violation, no?

> Note that a field is defined by referencing a specific RFC for Structured Fields, and the types available for use in its value are limited to those defined in that RFC. For example, a field whose definition references this document can have a value that uses the Date type ({{date}}), whereas a field whose definition references RFC 8941 cannot, because it will be treated as invalid (and likely discarded) by implementations of that specification.

Reviewer: ```suggestion
Note that a field is defined by referencing a specific RFC for Structured Fields, and the types available for use in its value are limited to those defined in that RFC. For example, a field whose definition references this document can have a value that uses the Date type ({{date}}), whereas a field whose definition references RFC 8941 cannot, because it will be treated as invalid (and therefore discarded) by implementations of that specification.
```

> Note that a field is defined by referencing a specific RFC for Structured Fields, and the types available for use in its value are limited to those defined in that RFC. For example, a field whose definition references this document can have a value that uses the Date type ({{date}}), whereas a field whose definition references RFC 8941 cannot, because it will be treated as invalid (and therefore discarded) by implementations of that specification.

Author: ```suggestion
Note that because the definition of a Strutured Field references a specific RFC for Structured Fields, the types available for use in its value are limited to those defined in that RFC. For example, a field whose definition references this document can have a value that uses the Date type ({{date}}), whereas a field whose definition references RFC 8941 cannot, because it will be treated as invalid (and therefore discarded) by implementations of that specification.

However, this document is designed to be backwards-compatible with RFC 8941; a parser that implements the requirements here can also parse valid Structured Fields whose definitions reference RFC 8941.
```

</artwork>
</blockquote>

Note that because the definition of a Strutured Field references a specific RFC for Structured Fields, the types available for use in its value are limited to those defined in that RFC. For example, a field whose definition references this document can have a value that uses the Date type ({{date}}), whereas a field whose definition references RFC 8941 cannot, because it will be treated as invalid (and therefore discarded) by implementations of that specification.

However, this document is designed to be backwards-compatible with RFC 8941; a parser that implements the requirements here can also parse valid Structured Fields whose definitions reference RFC 8941.

> In particular, the extensions available for use in a field are limited to the types allowed by that field's definition, which cannot exceed those defined by the Structured Fields specification it references.

Author: ```suggestion
Also note that this limitation also applies to future extensions to a field; for example, a field that is defined with reference to RFC 8941 cannot use the Date type, because some recipients might still be using an RFC 8941 parser to process it.
```

> However, this document is designed to be backwards-compatible with RFC 8941; a parser that implements the requirements here can also parse valid Structured Fields whose definitions reference RFC 8941.
> 
> Also note that this limitation also applies to future extensions to a field; for example, a field that is defined with reference to RFC 8941 cannot use the Date type, because some recipients might still be using an RFC 8941 parser to process it.

Author: ```suggestion
Also note that this limitation also applies to future extensions to a field; for example, a field that is defined with reference to RFC 8941 cannot use the Date type, because some recipients might still be using an RFC 8941 parser to process it.

However, this document is designed to be backwards-compatible with RFC 8941; a parser that implements the requirements here can also parse valid Structured Fields whose definitions reference RFC 8941.
```

> 

Author: ```suggestion
```

> 

Author: Hm, that's very wordy, and this section is already quite wordy.

> 

Reviewer 2: I know. I don’t like it much for that reason. But maybe this is worth the text. Slapping a section header on might help. “Backwards Compatibility” might be a good title.

> 

Reviewer 3: I cannot in good confidence say that any of this adds clarity, but I will not go as far as to call it surpluage, because there is a point to be made here.

I think what bugs me, is that having a SFbis parser return an unexpected and illegal sf-date is not materially different from having a 8941 parser return sf-decimal where sf-integer was expected.

In the end, the responsibility for validating the header can only fall on the application code.

Maybe we should simply write that, and use the 8941-defined-SFbis-parser-returning-sf-date example to illustrate a detail people might otherwise miss ?

> 

Author: See proposal below.

</artwork>
</blockquote>

> Note that because the definition of a Strutured Field references a specific RFC for Structured Fields, the types available for use in its value are limited to those defined in that RFC. For example, a field whose definition references this document can have a value that uses the Date type ({{date}}), whereas a field whose definition references RFC 8941 cannot, because it will be treated as invalid (and therefore discarded) by implementations of that specification.

Reviewer: ```suggestion
Note that because the definition of a Structured Field references a specific RFC for Structured Fields, the types available for use in its value are limited to those defined in that RFC. For example, a field whose definition references this document can have a value that uses the Date type ({{date}}), whereas a field whose definition references RFC 8941 cannot, because it will be treated as invalid (and therefore discarded) by implementations of that specification.
```

> Also note that this limitation also applies to future extensions to a field; for example, a field that is defined with reference to RFC 8941 cannot use the Date type, because some recipients might still be using an RFC 8941 parser to process it.

Reviewer: ```suggestion
Further note that this limitation also applies to future extensions to a field; for example, a field that is defined with reference to RFC 8941 cannot use the Date type, because some recipients might still be using an RFC 8941 parser to process it.
```

Author: I prefer also.

However, this document is designed to be backwards-compatible with RFC 8941; a parser that implements the requirements here can also parse valid Structured Fields whose definitions reference RFC 8941.

> The effect of upgrading a Structured Fields implementation is that some field values that were invalid according to RFC 8941 might become valid when processed.  For instance, though it would not be valid for a field defined against RFC 8941, a field might include include a Date value. An RFC 8941 implementation would reject the entire field, whereas an updated Structured Field implementation might permit the value. In some cases, the resulting Date value will be rejected by field-specific logic, but values in fields that are otherwise ignored (such as extension parameters) might not be detected and the field might subsequently be accepted and processed.

Author: ```suggestion
Because of that backwards compatibility, field values that are invalid according to RFC 8941 might become valid when parsed by an implementation of this specification.  For example, a field might be erroneously sent with a Date value. An RFC 8941 implementation would reject the entire field, whereas a parser written to this specification would accept it. Whether that would cause interoperability issues depends on whether the results of parsing are checked appropriately for a type mismatch.```

Reviewer 3: The "for example" grinds me, when that literally is the only possible thing that can happen.

Shorter and more to the point:

However, this document is designed to be backwards-compatible with RFC 8941; a parser that implements the requirements here can also parse valid Structured Fields whose definitions reference RFC 8941, but if the parser output contains any Date values, the caller must reject it as invalid, to remain RFC 8941 compliant.

Author: That's a bit _too_ short; see revision.

</artwork>
</blockquote>

> Note that a field is defined by referencing a specific RFC for Structured Fields, and its value is limited to the types defined in that RFC. For example, a field whose definition references this document can use the Date type ({{date}}), whereas a field whose definition references RFC8941 cannot. This also limits the kinds of extension values that can be later defined for a given field.

Reviewer 2: I might add:

> Otherwise, an extension that uses a type that is permitted by a newer version of the Structured Field specification might be treated as invalid (and discarded) by an implementation that only implements an older version.

I don't know if you want to say that there is a possibility that though an extension like this would be accepted by those implementations that support the newer version, that's not consistently interoperable.

Note also this this doesn't just apply to types.  It might apply if the parsing/validation rules for a type becomes more lenient in a subsequent version.

Author: See revision
